src/gui/programupdater.cpp

```diff
@@ -36,66 +37,95 @@
+            if (newVersion == currentVersion)
+            {
+                const bool isDevVersion = QString::fromLatin1(QBT_VERSION_STATUS).contains(
+                    QRegularExpression(QLatin1String("(alpha|beta|rc)")));
+                if (isDevVersion)
+                    return true;
+            }
+            return (newVersion > currentVersion);
+        }
+        catch (const std::runtime_error &)
+        {
+            return false;
+        }
+    }
 }
 
-ProgramUpdater::ProgramUpdater(QObject *parent, bool invokedByUser)
-    : QObject(parent)
-    , m_invokedByUser(invokedByUser)
+ProgramUpdater::ProgramUpdater(QObject *parent)
+    : QObject {parent}
 {
 }
 
-void ProgramUpdater::checkForUpdates()
+void ProgramUpdater::checkForUpdates() const
 {
+    const auto RSS_URL = QString::fromLatin1("https://www.fosshub.com/feed/5b8793a7f9ee5a5c3e97a3b2.xml");
     // Don't change this User-Agent. In case our updater goes haywire,
     // the filehost can identify it and contact us.
     Net::DownloadManager::instance()->download(
-                Net::DownloadRequest(RSS_URL).userAgent("qBittorrent/" QBT_VERSION_2 " ProgramUpdater (www.qbittorrent.org)")
-                , this, &ProgramUpdater::rssDownloadFinished);
+        Net::DownloadRequest(RSS_URL).userAgent("qBittorrent/" QBT_VERSION_2 " ProgramUpdater (www.qbittorrent.org)")
+        , this, &ProgramUpdater::rssDownloadFinished);
 }
 
-void ProgramUpdater::rssDownloadFinished(const Net::DownloadResult &result)
+QString ProgramUpdater::getNewVersion() const
 {
+    return m_newVersion;
+}
 
+void ProgramUpdater::rssDownloadFinished(const Net::DownloadResult &result)
+{
     if (result.status != Net::DownloadStatus::Success)
     {
         qDebug() << "Downloading the new qBittorrent updates RSS failed:" << result.errorString;
-        emit updateCheckFinished(false, QString(), m_invokedByUser);
+        emit updateCheckFinished();
         return;
     }
 
     qDebug("Finished downloading the new qBittorrent updates RSS");
 
+    const auto getStringValue = [](QXmlStreamReader &xml) -> QString
+    {
+        xml.readNext();
+        return (xml.isCharacters() && !xml.isWhitespace())
+            ? xml.text().toString()
+            : QString {};
+    };
+
 #ifdef Q_OS_MACOS
     const QString OS_TYPE {"Mac OS X"};
 #elif defined(Q_OS_WIN)
-    const QString OS_TYPE
-    {(::IsWindows7OrGreater()
-            && QSysInfo::currentCpuArchitecture().endsWith("64"))
+    const QString OS_TYPE {(::IsWindows7OrGreater()
+        && QSysInfo::currentCpuArchitecture().endsWith("64"))
         ? "Windows x64" : "Windows"};
 #endif
 
-    QString version;
-    QXmlStreamReader xml(result.data);
     bool inItem = false;
+    QString version;
     QString updateLink;
     QString type;
 
-    while (!xml.atEnd())
+    for (QXmlStreamReader xml(result.data); !xml.atEnd(); )
```

Comment:
Really dislike such declaration of `for` loop when some part is omitted (purely subjective).
Frustrated that they didn't add in new standard `while` loop with an initializer by anology with `if`.

Reply:
Seems you can avoid space before `)` (unless there is readability issue for you).

Reply:
I updated the `for` statement.

Reply:
>I updated the `for` statement.

Seems it is incorrect now...
You should read from stream before you process it. So previous code had `readNext()` at start of each iteration.

I can see that your intention was to restrict the scope of the variable, so you replaced `while` with `for`. Unfortunately, it seems impossible to achieve your goal using the full declaration of `for` loop. So either go back to your previous version, or to the `while` loop (I didn't see anything critically wrong with that).

Reply:
My previous comment about `for` loop was mostly related to the fact that often the use of an incomplete declaration is unjustified, because it can be easily replaced with a full one, or with another loop, for example, `while`.

Don't forget that you can always limit the scope just by using `{ }` around it.

Reply:
>You should read from stream before you process it. So previous code had readNext() at start of each iteration.

I noticed that and tested it, it works fine.

Reply:
It turns out that you process a nonexistent element in the first iteration, but skip the last element. Perhaps the algorithm in the loop body is just insensitive to this...

Reply:
>It turns out that you process a nonexistent element in the first iteration

yes, the stream is just initialized and nothing is there and the loop body is fine with it.

>but skip the last element

The last element before exiting the loop is still the same as before, nothing is skipped over.

Offtopic, I wonder why here used a streaming xml parser while complete xml data is present, I would imagine using another parser would yield more sensible code. Perhaps others would like to handle it...

Reply:
>> but skip the last element
>
>The last element before exiting the loop is still the same as before, nothing is skipped over.
>Before this PR, it seems readNext() will read into 1 additional item after the end of stream and this PR won't.

Forget it, I realized the semantics of `QXmlStreamReader::atEnd()` is unclear to me and I don't intend to figure it out, so I reverted it back to `while()` loop.